include/exadg/poisson/spatial_discretization/operator.cpp

```diff
@@ -453,6 +456,86 @@ Operator<dim, Number, n_components>::vmult_matrix_based(
 }
 #endif
 
+#ifdef DEAL_II_WITH_PETSC
+template<int dim, typename Number, int n_components>
+void
+Operator<dim, Number, n_components>::init_system_matrix(
+  PETScWrappers::MPI::SparseMatrix & system_matrix) const
+{
+  laplace_operator.init_system_matrix(system_matrix);
+}
+
+template<int dim, typename Number, int n_components>
+void
+Operator<dim, Number, n_components>::calculate_system_matrix(
+  PETScWrappers::MPI::SparseMatrix & system_matrix) const
+{
+  laplace_operator.calculate_system_matrix(system_matrix);
+}
+
+template<int dim, typename Number, int n_components>
+void
+Operator<dim, Number, n_components>::vmult_matrix_based(
+  VectorTypeDouble &                       dst,
+  PETScWrappers::MPI::SparseMatrix const & system_matrix,
+  VectorTypeDouble const &                 src) const
+{
+  // copy to petsc internal vector type because there is currently no such
+  // function in deal.II (and the transition via ReadWriteVector is too
+  // slow/poorly tested)
+  Vec vector_dst, vector_src;
+  VecCreateMPI(dst.get_mpi_communicator(),
+               dst.get_partitioner()->local_size(),
+               PETSC_DETERMINE,
+               &vector_dst);
+  VecCreateMPI(src.get_mpi_communicator(),
+               src.get_partitioner()->local_size(),
+               PETSC_DETERMINE,
+               &vector_src);
+
+  {
+    PetscInt       begin, end;
+    PetscErrorCode ierr = VecGetOwnershipRange(vector_src, &begin, &end);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+    PetscScalar * ptr;
+    ierr = VecGetArray(vector_src, &ptr);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+    const PetscInt local_size = src.get_partitioner()->local_size();
+    AssertDimension(local_size, static_cast<unsigned int>(end - begin));
+    for(PetscInt i = 0; i < local_size; ++i)
+    {
+      ptr[i] = src.local_element(i);
+    }
+
+    ierr = VecRestoreArray(vector_src, &ptr);
+  }
+
+  PETScWrappers::VectorBase petsc_dst(vector_dst);
+  system_matrix.vmult(petsc_dst, PETScWrappers::VectorBase(vector_src));
+
+  {
+    PetscInt       begin, end;
+    PetscErrorCode ierr = VecGetOwnershipRange(vector_dst, &begin, &end);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+    PetscScalar * ptr;
+    ierr = VecGetArray(vector_dst, &ptr);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+    const PetscInt local_size = dst.get_partitioner()->local_size();
+    AssertDimension(local_size, static_cast<unsigned int>(end - begin));
+    for(PetscInt i = 0; i < local_size; ++i)
+    {
+      dst.local_element(i) = ptr[i];
+    }
+
+    ierr = VecRestoreArray(vector_dst, &ptr);
+  }
+}
```

Comment:
Would it be possible to put this into a free function, say 

```suggestion
template<typename VectorTypeDealii>
void
apply_petsc_operation(VectorTypeDealii &dst, VectorTypeDealii const &src, lambda petsc_operation)
{
// pre copy

 petsc_operation(petsc_dst, petsc_src);

// post copy
}
```
It seems as if the same code is implemented in the AMG preconditioner.

Reply:
Yes, that is a good idea, see e2a3cd2e